lib/awful/keygrabber.lua

```diff
@@ -79,19 +661,80 @@ end
 --     end
 --   end)
 -- end
-function keygrabber.run(g)
+-- @function awful.keygrabber.run
+function keygrab.run(g)
     -- Remove the grabber if it is in the stack.
-    keygrabber.stop(g)
+    keygrab.stop(g)
+
     -- Record the grabber that has been added most recently.
     table.insert(grabbers, 1, g)
+
     -- Start the keygrabber if it is not running already.
     if not keygrabbing then
         keygrabbing = true
         capi.keygrabber.run(grabber)
     end
+
     return g
 end
 
-return keygrabber
+-- Implement the signal system for the keygrabber.
+
+local signals = {}
+
+--- Connect to a signal for all keygrabbers at once.
+-- @function awful.keygrabber.connect_signal
+-- @tparam string name The signal name.
+-- @tparam function callback The callback.
+function keygrab.connect_signal(name, callback)
+    signals[name] = signals[name] or {}
+
+    -- Use table.insert instead of signals[name][callback] = true to make
+    -- the execution order stable across `emit_signal`. This avoids some
+    -- heisenbugs.
+    table.insert(signals[name], callback)
+end
+
+--- Disconnect to a signal for all keygrabbers at once.
+-- @function awful.keygrabber.disconnect_signal
+-- @tparam string name The signal name.
+-- @tparam function callback The callback.
+function keygrab.disconnect_signal(name, callback)
+    signals[name] = signals[name] or {}
+
+    for k, v in ipairs(signals[name]) do
+        if v == callback then
+            table.remove(signals[name], k)
+            break
+        end
+    end
+end
+
+--- Emit a signal on the keygrabber module itself.
+--
+-- Warning, usually don't use this directly, use
+-- `my_keygrabber:emit_signal(name, ...)`. This function works on the whole
+-- keygrabber module, not one of its instance.
+--
+-- @function awful.keygrabber.emit_signal
+-- @tparam string name The signal name.
+-- @param ... Other arguments for the callbacks.
+function keygrab.emit_signal(name, ...)
+    signals[name] = signals[name] or {}
+
+    for _, cb in ipairs(signals[name]) do
+        cb(...)
+    end
+end
+
+function keygrab.get_is_running()
+    return keygrab.current_instance ~= nil
+end
+
+--@DOC_object_COMMON@
```

Comment:
Correct me if I'm wrong, but this line brings in another round of documentation for the signal APIs, plus it brings in `weak_connect_signal`, which you did not actually implement for your reimplementation.